programs/drift_vaults/src/state/tokenized_vault_depositor.rs

```diff
@@ -0,0 +1,574 @@
+        )?;
+
+        Ok(rebase_divisor)
+    }
+}
+
+impl TokenizedVaultDepositor {
+    pub fn new(vault: Pubkey, pubkey: Pubkey, mint: Pubkey, now: i64) -> Self {
+        Self {
+            vault,
+            pubkey,
+            mint,
+            vault_shares: 0,
+            last_vault_shares: 0,
+            last_valid_ts: now,
+            net_deposits: 0,
+            total_deposits: 0,
+            total_withdraws: 0,
+            cumulative_profit_share_amount: 0,
+            profit_share_fee_paid: 0,
+            vault_shares_base: 0,
+            bump: 0,
+            padding: [0; 3],
+        }
+    }
+
+    pub fn apply_profit_share(
+        self: &mut TokenizedVaultDepositor,
+        vault_equity: u64,
+        vault: &mut Vault,
+    ) -> Result<u64> {
+        let total_amount = depositor_shares_to_vault_amount(
+            self.get_vault_shares(),
+            vault.total_shares,
+            vault_equity,
+        )?;
+
+        let profit_share: u64 = self
+            .calculate_profit_share_and_update(total_amount, vault)?
+            .cast()?;
+
+        let profit_share_shares: u128 =
+            vault_amount_to_depositor_shares(profit_share, vault.total_shares, vault_equity)?;
+
+        self.decrease_vault_shares(profit_share_shares, vault)?;
+
+        vault.user_shares = vault.user_shares.safe_sub(profit_share_shares)?;
+
+        vault.manager_total_profit_share = vault
+            .manager_total_profit_share
+            .saturating_add(profit_share);
+
+        Ok(profit_share)
+    }
+
+    pub fn tokenize_shares(
+        self: &mut TokenizedVaultDepositor,
+        vault: &mut Vault,
+        vault_protocol: &mut Option<RefMut<VaultProtocol>>,
+        mint_supply: u64,
+        vault_equity: u64,
+        shares_transferred: u128,
+        now: i64,
+    ) -> Result<u64> {
+        let rebase_divisor = self.apply_rebase(vault, vault_protocol, vault_equity)?;
+        if rebase_divisor.is_some() {
+            return Err(ErrorCode::InvalidVaultRebase.into());
+        }
+
+        let shares_transferred = if rebase_divisor.is_some() {
+            shares_transferred
+                .checked_div(rebase_divisor.unwrap())
+                .expect("math")
+        } else {
+            shares_transferred
+        };
+
+        let VaultFee {
+            management_fee_payment,
+            management_fee_shares,
+            protocol_fee_payment: _protocol_fee_payment,
+            protocol_fee_shares: _protocol_fee_shares,
+        } = vault.apply_fee(vault_protocol, vault_equity, now)?;
+        let profit_share: u64 = self.apply_profit_share(vault_equity, vault)?;
+
+        let vault_shares_before = self.checked_vault_shares(vault)?;
+        let total_vault_shares_before = vault.total_shares;
+        let user_vault_shares_before = vault.user_shares;
+
+        let new_last_vault_shares = self
+            .last_vault_shares
+            .checked_add(shares_transferred)
```

Comment:
you should be able to use the safe_math library too? in prints out the line the math is hit, so useful for debugging